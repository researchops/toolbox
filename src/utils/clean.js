export function clean(data, ...field_names) {
	return data.map((item) =>
		field_names.reduce((acc, key) => {
			const field_value = item[key];
			let qualified = false;
			if (Array.isArray(field_value)) {
				qualified =
					field_value.length > 0 &&
					!field_value.every((item) => item.trim() === '');
			} else {
				qualified = field_value !== null;
			}
			if (qualified) {
				acc[key] = field_value;
			}
			return acc;
		}, {})
	)
}
