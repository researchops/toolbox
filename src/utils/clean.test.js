import { expect, test } from 'vitest';
import { clean } from './clean';

const test_data = [
	{
		expertise: ['User or design research'],
		experience: null,
		responsibilities: [
			'Generalist ',
			'Research libraries',
			'Recruitment',
			'Training and mentoring',
			'Finance and procurement',
			'Vendor management and tools',
			'Taxonomy',
			null,
			''
		]
	},
	{
		expertise: ['User or design research'],
		experience: '4 - 6 years',
		responsibilities: [
			'Training and mentoring',
			'Vendor management and tools',
			'',
		]
	},
	{
		expertise: ['User or design research'],
		experience: '',
		responsibilities: [
			'Training and mentoring',
			'Vendor management and tools',
			'Taxonomy'
		]
	}
];

test('remove empties / undefined from single choice charts', () => {
	const result = clean(test_data, 'experience');
	expect(result).toMatchSnapshot()
});

test('remove empties / undefined from multi-choice charts', () => {
	const result = clean(test_data, 'responsibilities');
	expect(result).toMatchSnapshot()
});

test('remove empties / undefined from multiple fields', () => {
	const result = clean(test_data, 'expertise', 'experience');
	expect(result).toMatchSnapshot()
});
