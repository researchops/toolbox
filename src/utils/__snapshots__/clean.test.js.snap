// Vitest Snapshot v1

exports[`remove empties / undefined from multi-choice charts 1`] = `
[
  {
    "responsibilities": [
      "Generalist ",
      "Research libraries",
      "Recruitment",
      "Training and mentoring",
      "Finance and procurement",
      "Vendor management and tools",
      "Taxonomy",
    ],
  },
  {
    "responsibilities": [
      "Training and mentoring",
      "Vendor management and tools",
    ],
  },
  {
    "responsibilities": [
      "Training and mentoring",
      "Vendor management and tools",
      "Taxonomy",
    ],
  },
]
`;

exports[`remove empties / undefined from multiple fields 1`] = `
[
  {
    "expertise": [
      "User or design research",
    ],
  },
  {
    "experience": "4 - 6 years",
    "expertise": [
      "User or design research",
    ],
  },
  {
    "expertise": [
      "User or design research",
    ],
  },
]
`;

exports[`remove empties / undefined from single choice charts 1`] = `
[
  {
    "experience": "4 - 6 years",
  },
]
`;
